java/server/src/co/mitro/core/servlets/GetMyPrivateKey.java

```diff
@@ -221,13 +227,43 @@ private void sendEmailAndThrow(GetMyPrivateKeyRequest in, MitroRequestContext co
       String token = makeLoginTokenString(identity, in.extensionId, in.deviceId);
       String tokenSignature = TwoFactorSigningService.signToken(token);
 
-      if (!context.manager.isReadOnly()) {
-        DBEmailQueue email = DBEmailQueue.makeNewDeviceVerification(identity.getName(), token, tokenSignature, context.platform, context.requestServerUrl);
-        context.manager.emailDao.create(email);
-        context.manager.commitTransaction();
+      GenericRawResults<String[]> queuedEmails = context.manager.emailDao.queryRaw(
+              "SELECT type_string, arg_string FROM ("
+              + "(SELECT type_string, arg_string FROM email_queue)"
+              + " UNION "
+              + "(SELECT type_string, arg_string FROM email_queue_sent) "
+              + ") WHERE type_string=? AND arg_string LIKE ?",
+              DBEmailQueue.Type.LOGIN_ON_NEW_DEVICE.getValue(),
+              "%" + identity.getName() + "%"
```

Comment:
what's the origin of identity? this seems dangerous, e.g. providing an empty like requires a full table scan (cheap in our case but could be used as DOS). Also not sure we should check in `email_queue_sent` because sending may have failed for another reason. So we should allow people to trigger another send. Unless there's a new timestamp and hence new token every time?